docs/architecture/adr-059-go-api-stability.md

```diff
@@ -0,0 +1,193 @@
+
+* **Private:** a Go identifier starting with a lowercase letter, which therefore cannot be accessed by an external package unless via an exported field, variable, or function/method return value.
+
+* **Public API:** any Go identifier that can be imported or accessed by an external project, except test code in `_test.go` files.
+
+* **Private API:** any Go identifier that is not accessible via a public API, including all code in the internal directory.
+
+## Alternative Approaches
+
+- Split all public APIs out to separate Go modules in separate Git repositories, and consider all Tendermint code internal and not subject to API backwards compatibility at all. This was rejected, since it has been attempted by the Tendermint project earlier, resulting in too much dependency management overhead.
+
+- Simply document which APIs are public and which are private. This is the current approach, but users should not be expected to self-enforce this, and external projects will often end up depending on internal code anyway.
+
+## Decision
+
+From Tendermint 1.0, all internal code (except private APIs) will be placed in a root-level [`internal` directory](https://golang.org/cmd/go/#hdr-Internal_Directories), which the Go compiler will block for use by external projects. All exported items outside of the `internal` directory are considered a public API and subject to backwards compatibility guarantees, except files ending in `_test.go`.
+
+The `crypto` package will be split out to a separate module in a separate repo. This is the main general-purpose package used by external projects, and is the only Tendermint dependency in e.g. IAVL which can cause some problems for projects depending on both IAVL and Tendermint.
+
+The `tm-db` package will remain a separate module in a separate repo.
+
+## Detailed Design
+
+### Public API
+
+When preparing our public API for 1.0, we should keep these principles in mind:
+
+- Limit the number of public APIs that we start out with - we can always add new APIs later, but we can't change or remove APIs once they're made public.
+
+- Before an API is made public, do a thorough review of the API to make sure it covers any future needs, can accomodate expected changes, and follows good API design practices.
+
+The following current packages will be part of our public API in 1.0 in some form:
+
+- `abci`
+- `config`
+- `crypto` (separate module)
+- `libs/log`
+- `light`
+- `node`
+- `privval` (client parts)
+- `rpc/client`
+- `types` (respective packages)
+- `version`
+
+For comparison, the following are the number of Tendermint imports in the Cosmos SDK (excluding tests) - anything not listed above must be removed from the SDK or implemented there:
+
+```
+      1 github.com/tendermint/tendermint/abci/server
+     73 github.com/tendermint/tendermint/abci/types
+      2 github.com/tendermint/tendermint/cmd/tendermint/commands
```

Comment:
Some applications import and wrap the some of the cmds. Are there other ways applications will wrap Tendermint cmds?

Reply:
Yeah, I'm not really familiar with this bit, but would like to know more. Why would they wrap the commands rather than instantiate a `Node` and start it?

Reply:
They do instantiate a node via `NewNode`(https://github.com/cosmos/cosmos-sdk/blob/4faeefebd2abc50f89b02daf836f2edb77f61288/server/start.go#L214-L214). But in some cases like: https://github.com/cosmos/cosmos-sdk/blob/64b6bb5270e1a3b688c2d98a8f481ae04bb713ca/server/tm_cmds.go#L154-L154 a user will call `simd unsafe-reset-all` to reset the Tendermint data directory. Currently, I can not think of a way to implement this on the application side. 

Also, part of the private privval: should be exposed to allow users to see their node-id: https://github.com/cosmos/cosmos-sdk/blob/64b6bb5270e1a3b688c2d98a8f481ae04bb713ca/server/tm_cmds.go#L51-L51, this may be possible to be implemented on the application side. The nodekey, is another area that may need to be exposed: https://github.com/cosmos/cosmos-sdk/blob/64b6bb5270e1a3b688c2d98a8f481ae04bb713ca/server/tm_cmds.go#L31-L31

Reply:
Making the command-line tools public is the sort of thing that will pull almost anything else in the code base into the public API as well, since they depend on everything else. I also feel like user interfaces is something that should be application-specific, and not coupled too tightly to Tendermint.

Stuff like `tcmd.ResetAll` should probably be a method on `Node` or similar. The `command` package should only be concerned with the actual user interface, not with the functionality it exposes.

As for privval, we can discuss the details of which bits must be public once we start splitting it up. Functionality that users _need_ must be exposed to them somehow.